// @webpd/runtime/dist/src/fs-web/fake-filesystem.js
import { audioBufferToArray, fixSoundChannelCount, fetchFile } from '../utils.js';

/*
 * Copyright (c) 2022-2023 Sébastien Piquemal <sebpiq@protonmail.com>, Chris McCormick.
 *
 * This file is part of WebPd
 * (see https://github.com/sebpiq/WebPd).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
const FILES = {};
const STREAMS = {};
class FakeStream {
    constructor(url, sound) {
        this.url = url;
        this.sound = sound;
        this.frameCount = sound[0].length;
        this.readPosition = 0;
    }
}
const read = async (url) => {
    if (FILES[url]) {
        return FILES[url];
    }
    const arrayBuffer = await fetchFile(url);
    return {
        type: 'binary',
        data: arrayBuffer,
    };
};
// TODO : testing
const readSound = async (url, context) => {
    let fakeFile = FILES[url] || (await read(url));
    switch (fakeFile.type) {
        case 'binary':
            const audioBuffer = await context.decodeAudioData(fakeFile.data);
            return audioBufferToArray(audioBuffer);
        case 'sound':
            // We copy the data here o it can be manipulated freely by the host.
            // e.g. if the buffer is sent as transferrable to the node we don't want the original to be transferred.
            return fakeFile.data.map((array) => array.slice());
    }
};
const writeSound = async (sound, url) => {
    FILES[url] = {
        type: 'sound',
        data: sound,
    };
};
const readStreamSound = async (operationId, url, channelCount, context) => {
    const sound = await readSound(url, context);
    STREAMS[operationId] = new FakeStream(url, fixSoundChannelCount(sound, channelCount));
    return STREAMS[operationId];
};
const writeStreamSound = async (operationId, url, channelCount) => {
    const emptySound = [];
    for (let channel = 0; channel < channelCount; channel++) {
        emptySound.push(new Float32Array(0));
    }
    STREAMS[operationId] = new FakeStream(url, emptySound);
    FILES[url] = {
        type: 'sound',
        data: emptySound,
    };
    return STREAMS[operationId];
};
const getStream = (operationId) => {
    return STREAMS[operationId];
};
const killStream = (operationId) => {
    console.log('KILL STREAM', operationId);
    delete STREAMS[operationId];
};
const pullBlock = (stream, frameCount) => {
    const block = stream.sound.map((array) => array.slice(stream.readPosition, stream.readPosition + frameCount));
    stream.readPosition += frameCount;
    return block;
};
const pushBlock = (stream, block) => {
    stream.sound = stream.sound.map((channelData, channel) => {
        const concatenated = new Float32Array(channelData.length + block[channel].length);
        concatenated.set(channelData);
        concatenated.set(block[channel], channelData.length);
        return concatenated;
    });
    stream.frameCount = stream.sound[0].length;
    FILES[stream.url].data = stream.sound;
};
var fakeFs = {
    writeSound,
    readSound,
    readStreamSound,
    writeStreamSound,
    pullBlock,
    pushBlock,
};

export { FakeStream, fakeFs as default, getStream, killStream, pullBlock, pushBlock, readSound };
