// @webpd/runtime/dist/src/utils.js
import fetchRetry$1 from '../node_modules/fetch-retry/index.js';

/*
 * Copyright (c) 2022-2023 Sébastien Piquemal <sebpiq@protonmail.com>, Chris McCormick.
 *
 * This file is part of WebPd
 * (see https://github.com/sebpiq/WebPd).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
const fetchRetry = fetchRetry$1(fetch);
/**
 * Note : the audio worklet feature is available only in secure context.
 * This function will fail when used in insecure context (non-https, etc ...)
 * @see https://developer.mozilla.org/en-US/docs/Web/API/AudioWorklet
 */
const addModule = async (context, processorCode) => {
    const blob = new Blob([processorCode], { type: 'text/javascript' });
    const workletProcessorUrl = URL.createObjectURL(blob);
    return context.audioWorklet.addModule(workletProcessorUrl);
};
// TODO : testing
const fetchFile = async (url) => {
    let response;
    try {
        response = await fetchRetry(url, { retries: 3 });
    }
    catch (err) {
        throw new FileError(response.status, err.toString());
    }
    if (!response.ok) {
        const responseText = await response.text();
        throw new FileError(response.status, responseText);
    }
    return response.arrayBuffer();
};
const audioBufferToArray = (audioBuffer) => {
    const sound = [];
    for (let channel = 0; channel < audioBuffer.numberOfChannels; channel++) {
        sound.push(audioBuffer.getChannelData(channel));
    }
    return sound;
};
// TODO : testing
const fixSoundChannelCount = (sound, targetChannelCount) => {
    if (sound.length === 0) {
        throw new Error(`Received empty sound`);
    }
    const floatArrayType = sound[0].constructor;
    const frameCount = sound[0].length;
    const fixedSound = sound.slice(0, targetChannelCount);
    while (sound.length < targetChannelCount) {
        fixedSound.push(new floatArrayType(frameCount));
    }
    return fixedSound;
};
const resolveRelativeUrl = (rootUrl, relativeUrl) => {
    return new URL(relativeUrl, rootUrl).href;
};
class FileError extends Error {
    constructor(status, msg) {
        super(`Error ${status} : ${msg}`);
    }
}

export { FileError, addModule, audioBufferToArray, fetchFile, fixSoundChannelCount, resolveRelativeUrl };
