// @webpd/runtime/dist/src/WebPdWorkletNode.js
import WEB_PD_WORKLET_PROCESSOR_CODE from './assets/WebPdWorkletProcessor.js.txt.js';
import ASSEMBLY_SCRIPT_WASM_BINDINGS_CODE from '../node_modules/@webpd/compiler/dist/assemblyscript-wasm-bindings.iife.js';
import JAVA_SCRIPT_BINDINGS_CODE from '../node_modules/@webpd/compiler/dist/javascript-bindings.iife.js';
import { addModule } from './utils.js';

/*
 * Copyright (c) 2022-2023 Sébastien Piquemal <sebpiq@protonmail.com>, Chris McCormick.
 *
 * This file is part of WebPd
 * (see https://github.com/sebpiq/WebPd).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
// TODO : manage transferables
class WebPdWorkletNode extends AudioWorkletNode {
    constructor(context) {
        super(context, 'webpd-node', {
            numberOfOutputs: 1,
            outputChannelCount: [2],
        });
    }
    destroy() {
        this.port.postMessage({
            type: 'destroy',
            payload: {},
        });
    }
}
// Concatenate WorkletProcessor code with the Wasm bindings it needs
const WEBPD_WORKLET_PROCESSOR_CODE = ASSEMBLY_SCRIPT_WASM_BINDINGS_CODE +
    ';\n' +
    JAVA_SCRIPT_BINDINGS_CODE +
    ';\n' +
    WEB_PD_WORKLET_PROCESSOR_CODE;
const registerWebPdWorkletNode = (context) => {
    return addModule(context, WEBPD_WORKLET_PROCESSOR_CODE);
};

export { WebPdWorkletNode as default, registerWebPdWorkletNode };
